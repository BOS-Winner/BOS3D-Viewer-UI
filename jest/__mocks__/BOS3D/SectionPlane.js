/* eslint-disable lines-between-class-members */
class SectionPlane {
  constructor() {
    this.coordinateSystem = {
      hide: () => {},
    };
  }

  setDirection() {}
  setPlane() {}
  hidePlane() {}
  showPlane() {}
  exit() {}
}

export default SectionPlane;
