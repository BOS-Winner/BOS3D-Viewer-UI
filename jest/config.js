module.exports = {
  rootDir: '../',
  testMatch: [
    "**/__tests__/**/*.test.{js,jsx}",
  ],
  moduleNameMapper: {
    "\\.less$": "identity-obj-proxy",
    "\\.css$": "<rootDir>/jest/__mocks__/styleMock.js",
    "Libs(.*)$": "<rootDir>/src/Libs$1",
    "UIutils(.*)$": "<rootDir>/src/UI/utils$1",
    "^Base(.*)$": "<rootDir>/src/UI/Base$1",
    "^BOS3D(.*)$": "<rootDir>/jest/__mocks__/BOS3D$1",
    "^BOS2D$": "<rootDir>/jest/__mocks__/BOS2D.js",
    "^BOS3DUI$": "<rootDir>/jest/__mocks__/BOS3DUI.js",
    "^jest/(.*)$": "<rootDir>/jest/$1",
    "^mock/(.*)$": "<rootDir>/jest/__mocks__/$1",
  },
  moduleFileExtensions: ["js", "jsx"],
  moduleDirectories: ["node_modules"],
  transform: {
    "\\.(js|jsx)": "babel-jest",
    "\\.(jpg|jpeg|png|gif|svg)$": "<rootDir>/jest/__mocks__/fileMock.js",
  },
  transformIgnorePatterns: [
    // Change MODULE_NAME_HERE to your module that isn't being compiled
    // "/node_modules/(?!three|lodash-es).+\\.js$"
    "/node_modules/core-js",
    "/node_modules/@testing-library",
    "/node_modules/@babel",
  ],
  coveragePathIgnorePatterns: [
    "/node_modules/",
    "/jest/"
  ],
};
