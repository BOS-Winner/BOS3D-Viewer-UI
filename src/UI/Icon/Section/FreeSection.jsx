import React from "react";
import ReactDOM from "react-dom";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import Icon from "Base/Icon";
import Modal from "Base/Modal";
import { changeMode } from "../action";
import toastr from "../../toastr";
import style from "./style.less";
import fuckIE from "../../theme/fuckIE.less";
import showPng from "../img/white/show@2x.png";
import hidePng from "../img/white/hide@2x.png";
import FreeSectionPng from "../img/white/freeSection.png";

class FreeSection extends React.Component {
  constructor(props) {
    super(props);
    this.sectionPlane = {};
    this.minWidth = props.viewer.viewportDiv.clientWidth < 760 ? 120 : 170;
    this.isTouchDevice = props.BIMWINNER.BOS3D.DeviceTest.isTouchDevice();
    this.touchProgress = 50;
    this.state = {
      dialogVisible: false,
      showPlane: true,
    };
  }

  componentDidUpdate(prevProps) {
    // 开启自由剖切
    if (prevProps.mode !== "自由剖切模式" && this.props.mode === "自由剖切模式") {
      if (prevProps.mode !== '') {
        toastr.info(`自由剖切模式下将关闭${prevProps.mode}`, "", {
          target: `#${this.props.viewer.viewport}`
        });
      }
      this.startFreeSection();
      this.props.onEnter();
    }
    // 关闭自由剖切
    if (prevProps.mode === "自由剖切模式" && this.props.mode !== "自由剖切模式") {
      this.sectionPlane.exit();
    }
  }

  changeDialog(visible) {
    if (!visible) {
      this.props.changeMode('');
    }
    this.setState({
      dialogVisible: visible
    });
  }

  onClick() {
    const selected = (this.props.mode === "自由剖切模式");
    this.setState(() => {
      this.props.changeMode(selected ? "" : "自由剖切模式");
      return {
        dialogVisible: !selected,
      };
    });
  }

  startFreeSection() {
    const BIMWINNER = this.props.BIMWINNER;
    this.sectionPlane = new BIMWINNER.BOS3D.SectionPlane({
      direction: "Forward",
      id: "SectionPlane",
      plane: "X",
      progress: 50,
      viewer: this.props.viewer
    });
    if (this.isTouchDevice) {
      this.sectionPlane.coordinateSystem.hide(true);
    }
    this.props.viewer.render();
  }

  setSectionPlane(e) {
    e.stopPropagation();
    const type = e.target.value;
    if (type[0] === '-') {
      this.sectionPlane.setDirection("Reverse");
      this.sectionPlane.setPlane(type[1]);
    } else {
      this.sectionPlane.setDirection("Forward");
      this.sectionPlane.setPlane(type);
    }
    if (this.isTouchDevice) {
      this.sectionPlane.setProgress(this.touchProgress);
    }
    this.props.viewer.render();
  }

  switchShowPlane(e) {
    e.preventDefault();
    e.stopPropagation();
    this.setState(state => {
      if (state.showPlane) {
        this.sectionPlane.hidePlane();
      } else {
        this.sectionPlane.showPlane();
      }
      return {
        showPlane: !state.showPlane,
      };
    });
  }

  setProgress(e) {
    e.stopPropagation();
    this.touchProgress = Number(e.target.value);
    this.sectionPlane.setProgress(this.touchProgress);
    this.props.viewer.render();
  }

  render() {
    const selected = (this.props.mode === "自由剖切模式");
    const imgTitle = this.state.showPlane ? "点击隐藏剖切面" : "点击显示剖切面";

    const selectDirection = (
      <select
        title="改变轴向"
        defaultValue="X"
        onChange={e => {
          this.setSectionPlane(e);
        }}
      >
        <option value="X">X</option>
        <option value="Y">Y</option>
        <option value="Z">Z</option>
        <option value="-X">-X</option>
        <option value="-Y">-Y</option>
        <option value="-Z">-Z</option>
      </select>
    );
    const showPlane = (
      <img
        className={style.eye}
        src={this.state.showPlane ? showPng : hidePng}
        alt={imgTitle}
        title={imgTitle}
        onClick={e => {
          this.switchShowPlane(e);
        }}
      />
    );

    let content = <></>;
    if (this.isTouchDevice) {
      content = ReactDOM.createPortal((
        <div
          role="presentation"
          className={style.touchBar}
          style={{
            display: this.state.dialogVisible && selected ? "flex" : "none"
          }}
          onClick={e => {
            e.preventDefault();
            e.stopPropagation();
          }}
        >
          {selectDirection}
          {showPlane}
          <span>0</span>
          <div className={style.range}>
            <input
              type="range"
              min={0}
              max={100}
              step={1}
              defaultValue={this.touchProgress}
              onChange={e => { this.setProgress(e) }}
            />
          </div>
          <span>100</span>
          <button
            type="button"
            onClick={e => {
              e.preventDefault();
              e.stopPropagation();
              this.changeDialog(false);
            }}
          >
            退出
          </button>
        </div>
      ), this.props.viewer.viewportDiv);
    } else {
      content = (
        <Modal
          visible={this.state.dialogVisible && selected}
          onCancel={() => {
            this.changeDialog(false);
          }}
          title="自由剖切选项"
          width={`${this.minWidth}px`}
          height="110px"
          destroyOnClose
          minWidth={this.minWidth}
          minHeight={100}
          viewportDiv={this.props.viewer.viewportDiv}
        >
          <div className={style.freeSection}>
            {selectDirection}
            {showPlane}
          </div>
        </Modal>
      );
    }
    return (
      <div
        className={selected ? fuckIE.select : ''}
        title="自由剖切"
        role="button"
        tabIndex={0}
        onClick={() => {
          this.onClick();
        }}
      >
        <Icon img={FreeSectionPng} />
        {content}
      </div>
    );
  }
}

FreeSection.propTypes = {
  changeMode: PropTypes.func.isRequired,
  mode: PropTypes.string.isRequired,
  onEnter: PropTypes.func.isRequired,
  viewer: PropTypes.object.isRequired,
  BIMWINNER: PropTypes.object.isRequired,
};

const mapStateToProps = (state) => ({
  mode: state.button.mode,
  viewer: state.system.viewer3D,
  BIMWINNER: state.system.BIMWINNER,
});
const mapDispatchToProps = (dispatch) => ({
  changeMode: mode => {
    dispatch(changeMode(mode));
  }
});
const WrappedContainer = connect(
  mapStateToProps,
  mapDispatchToProps
)(FreeSection);

export default WrappedContainer;
