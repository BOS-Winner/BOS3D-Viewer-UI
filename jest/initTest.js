import EventEmitter from "events";
import { createStore } from "redux";
import BOS3D from 'BOS3D';
import reducer from "../src/UI/reducer";

export function genStore() {
  const ee = new EventEmitter();
  const viewer3D = new BOS3D.Viewer();

  return createStore(reducer, {
    system: {
      viewer3D,
      BIMWINNER: { BOS3D },
      apiVersion: 'api',
      eventEmitter: ee,
    }
  });
}
