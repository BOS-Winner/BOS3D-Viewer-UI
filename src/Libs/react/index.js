import React from "react";
import useUpdateEffect from "./useUpdateEffect";

React.useUpdateEffect = useUpdateEffect;

export default React;
// export { default as useUpdateEffect } from "./useUpdateEffect";
