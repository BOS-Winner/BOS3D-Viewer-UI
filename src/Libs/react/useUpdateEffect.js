import { useEffect, useRef } from "react";

export default function (effect, deps) {
  const isFirstRender = useRef(true);
  useEffect(() => {
    if (isFirstRender.current) {
      isFirstRender.current = false;
    } else {
      effect();
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, deps);
}
