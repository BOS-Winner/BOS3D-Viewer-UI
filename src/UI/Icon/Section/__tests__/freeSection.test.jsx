/* eslint-disable camelcase */
/* eslint-disable react/prop-types */
import React from 'react';
import { Provider } from 'react-redux';
import { genStore } from 'jest/initTest';
import { render, act, fireEvent } from "@testing-library/react";
import SectionPlane from 'BOS3D/SectionPlane';
import FreeSection from '../FreeSection';

jest.mock('Base/Modal', () => props => (
  <div
    className={props.className}
    style={props.style}
  >
    {props.children}
  </div>
));
jest.mock('BOS3D/SectionPlane');
const onEnter = jest.fn();
let store;
let g_getByTitle;
let section;

describe('freeSection', () => {
  beforeEach(() => {
    store = genStore();
    const { container, getByTitle } = render((
      <Provider store={store}>
        <FreeSection onEnter={onEnter} />
      </Provider>
    ));
    section = container;
    g_getByTitle = getByTitle;
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should switch mode', () => {
    expect(section).toMatchSnapshot();
    act(() => {
      g_getByTitle('自由剖切').click();
    });
    expect(store.getState().button.mode).toEqual('自由剖切模式');
    expect(section).toMatchSnapshot();
    expect(onEnter).toBeCalledTimes(1);
    act(() => {
      g_getByTitle('自由剖切').click();
    });
    expect(store.getState().button.mode).toEqual('');
    expect(section).toMatchSnapshot();
  });

  it('should change axis direction', () => {
    const select = g_getByTitle('改变轴向');
    act(() => {
      g_getByTitle('自由剖切').click();
    });
    const sectionPlane = SectionPlane.mock.instances[0];
    fireEvent.change(select, {
      target: {
        value: 'Y'
      }
    });
    expect(sectionPlane.setDirection).lastCalledWith('Forward');
    expect(sectionPlane.setPlane).lastCalledWith('Y');
    fireEvent.change(select, {
      target: {
        value: '-Z'
      }
    });
    expect(sectionPlane.setDirection).lastCalledWith('Reverse');
    expect(sectionPlane.setPlane).lastCalledWith('Z');
  });

  it('should hide or show section face', () => {
    act(() => {
      g_getByTitle('自由剖切').click();
    });
    const sectionPlane = SectionPlane.mock.instances[0];
    const switchSectionFace = g_getByTitle('点击隐藏剖切面');
    act(() => {
      switchSectionFace.click();
    });
    expect(switchSectionFace.getAttribute('title')).toEqual('点击显示剖切面');
    expect(switchSectionFace.getAttribute('alt')).toEqual('点击显示剖切面');
    expect(switchSectionFace.getAttribute('src')).toEqual('hide@2x.png');
    expect(sectionPlane.hidePlane).toBeCalledTimes(1);
    act(() => {
      switchSectionFace.click();
    });
    expect(switchSectionFace.getAttribute('title')).toEqual('点击隐藏剖切面');
    expect(switchSectionFace.getAttribute('alt')).toEqual('点击隐藏剖切面');
    expect(switchSectionFace.getAttribute('src')).toEqual('show@2x.png');
    expect(sectionPlane.showPlane).toBeCalledTimes(1);
  });
});
