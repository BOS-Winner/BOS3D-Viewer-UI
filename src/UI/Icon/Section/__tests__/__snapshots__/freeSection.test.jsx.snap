// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`freeSection should switch mode 1`] = `
<div>
  <div
    class=""
    role="button"
    tabindex="0"
    title="自由剖切"
  >
    <div
      class="icon "
      data-selected="false"
      role="button"
      style="background-image: url(freeSection.png);"
      tabindex="0"
    />
    <div>
      <div
        class="freeSection"
      >
        <select
          title="改变轴向"
        >
          <option
            selected=""
            value="X"
          >
            X
          </option>
          <option
            value="Y"
          >
            Y
          </option>
          <option
            value="Z"
          >
            Z
          </option>
          <option
            value="-X"
          >
            -X
          </option>
          <option
            value="-Y"
          >
            -Y
          </option>
          <option
            value="-Z"
          >
            -Z
          </option>
        </select>
        <img
          alt="点击隐藏剖切面"
          class="eye"
          src="show@2x.png"
          title="点击隐藏剖切面"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`freeSection should switch mode 2`] = `
<div>
  <div
    class="select"
    role="button"
    tabindex="0"
    title="自由剖切"
  >
    <div
      class="icon "
      data-selected="false"
      role="button"
      style="background-image: url(freeSection.png);"
      tabindex="0"
    />
    <div>
      <div
        class="freeSection"
      >
        <select
          title="改变轴向"
        >
          <option
            selected=""
            value="X"
          >
            X
          </option>
          <option
            value="Y"
          >
            Y
          </option>
          <option
            value="Z"
          >
            Z
          </option>
          <option
            value="-X"
          >
            -X
          </option>
          <option
            value="-Y"
          >
            -Y
          </option>
          <option
            value="-Z"
          >
            -Z
          </option>
        </select>
        <img
          alt="点击隐藏剖切面"
          class="eye"
          src="show@2x.png"
          title="点击隐藏剖切面"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`freeSection should switch mode 3`] = `
<div>
  <div
    class=""
    role="button"
    tabindex="0"
    title="自由剖切"
  >
    <div
      class="icon "
      data-selected="false"
      role="button"
      style="background-image: url(freeSection.png);"
      tabindex="0"
    />
    <div>
      <div
        class="freeSection"
      >
        <select
          title="改变轴向"
        >
          <option
            selected=""
            value="X"
          >
            X
          </option>
          <option
            value="Y"
          >
            Y
          </option>
          <option
            value="Z"
          >
            Z
          </option>
          <option
            value="-X"
          >
            -X
          </option>
          <option
            value="-Y"
          >
            -Y
          </option>
          <option
            value="-Z"
          >
            -Z
          </option>
        </select>
        <img
          alt="点击隐藏剖切面"
          class="eye"
          src="show@2x.png"
          title="点击隐藏剖切面"
        />
      </div>
    </div>
  </div>
</div>
`;
